File: samples/snippets/visualbasic/System.IO/FileStreamOptions/PreallocationSizeExample.vb
'<snippet1>
Imports System.IO

Module PreallocationSizeExample
    Private Const SourcePath As String = "source.txt"
    Private Const DestinationPath As String = "destination.txt"

    Sub Main()
        CheckFiles()
        PreallocateAndCopy()

        'Destination contents should be the same as source contents
        Console.WriteLine(File.ReadAllText(DestinationPath))
    End Sub

    Private Sub PreallocateAndCopy()
        Dim openForReading = New FileStreamOptions With {
            .Mode = FileMode.Open
        }
        Using source = New FileStream(SourcePath, openForReading)

            Dim createForWriting = New FileStreamOptions With {
                .Mode = FileMode.CreateNew,
                .Access = FileAccess.Write,
                .PreallocationSize = source.Length ' specify size up-front
            }

            Using destination = New FileStream(DestinationPath, createForWriting)
                source.CopyTo(destination)
            End Using

        End Using

    End Sub

    Private Sub CheckFiles()
        If Not File.Exists(SourcePath) Then
            File.WriteAllText(SourcePath, "Hello world!")
        End If

        If File.Exists(DestinationPath) Then
            File.Delete(DestinationPath)
        End If
    End Sub

End Module
'</snippet1>

File: samples/snippets/csharp/System.IO/FileStreamOptions/PreallocationSizeExample.cs
//<snippet1>
using System;
using System.IO;

public static class PreallocationSizeExample
{
    private const string SourcePath = "source.txt";
    private const string DestinationPath = "destination.txt";

    public static void Main()
    {
        CheckFiles();
        PreallocateAndCopy();

        // Destination contents should be the same as source contents
        Console.WriteLine(File.ReadAllText(DestinationPath));
    }

    private static void PreallocateAndCopy()
    {
        var openForReading = new FileStreamOptions { Mode = FileMode.Open };
        using var source = new FileStream(SourcePath, openForReading);

        var createForWriting = new FileStreamOptions
        {
            Mode = FileMode.CreateNew,
            Access = FileAccess.Write,
            PreallocationSize = source.Length // specify size up-front
        };
        using var destination = new FileStream(DestinationPath, createForWriting);

        source.CopyTo(destination);
    }

    private static void CheckFiles()
    {
        if (!File.Exists(SourcePath))
        {
            File.WriteAllText(SourcePath, "Hello world!");
        }

        if (File.Exists(DestinationPath))
        {
            File.Delete(DestinationPath);
        }
    }
}
//</snippet1>
